metadata/src/lib.rs

```diff
@@ -3,844 +3,120 @@
-        );
-
-        #[allow(dead_code)]
-        #[derive(scale_info::TypeInfo)]
-        enum EnumFieldNotNamed {
-            First(u8),
-        }
-        #[allow(dead_code)]
-        #[derive(scale_info::TypeInfo)]
-        enum EnumFieldNotNamedSecond {
-            Second(u8),
-        }
-        // The enums are binary compatible, they contain a different semantic meaning:
-        // `First(u8)` and `Second(u8)`.
-        assert_ne!(
-            to_hash(meta_type::<EnumFieldNotNamed>()),
-            to_hash(meta_type::<EnumFieldNotNamedSecond>())
-        );
-
-        #[allow(dead_code)]
-        #[derive(scale_info::TypeInfo)]
-        enum EnumFieldNamed {
-            First { a: u8 },
-        }
-        #[allow(dead_code)]
-        #[derive(scale_info::TypeInfo)]
-        enum EnumFieldNamedSecond {
-            First { b: u8 },
-        }
-        // Named fields contain a different semantic meaning ('a' and 'b').
-        assert_ne!(
-            to_hash(meta_type::<EnumFieldNamed>()),
-            to_hash(meta_type::<EnumFieldNamedSecond>())
-        );
-
-        #[allow(dead_code)]
-        #[derive(scale_info::TypeInfo)]
-        struct StructFieldNamed {
-            a: u32,
-        }
-        #[allow(dead_code)]
-        #[derive(scale_info::TypeInfo)]
-        struct StructFieldNamedSecond {
-            b: u32,
-        }
-        // Similar to enums, struct fields contain a different semantic meaning ('a' and 'b').
-        assert_ne!(
-            to_hash(meta_type::<StructFieldNamed>()),
-            to_hash(meta_type::<StructFieldNamedSecond>())
-        );
-
-        #[allow(dead_code)]
-        #[derive(scale_info::TypeInfo)]
-        enum EnumField {
-            First,
-            // Field is unnamed, but has type name `u8`.
-            Second(u8),
-            // File is named and has type name `u8`.
-            Third { named: u8 },
-        }
-
-        #[allow(dead_code)]
-        #[derive(scale_info::TypeInfo)]
-        enum EnumFieldSwap {
-            Second(u8),
-            First,
-            Third { named: u8 },
-        }
-        // Swapping the registration order should also be taken into account.
-        assert_ne!(
-            to_hash(meta_type::<EnumField>()),
-            to_hash(meta_type::<EnumFieldSwap>())
-        );
-
-        #[allow(dead_code)]
-        #[derive(scale_info::TypeInfo)]
-        struct StructField {
-            a: u32,
-            b: u32,
-        }
-
-        #[allow(dead_code)]
-        #[derive(scale_info::TypeInfo)]
-        struct StructFieldSwap {
-            b: u32,
-            a: u32,
-        }
-        assert_ne!(
-            to_hash(meta_type::<StructField>()),
-            to_hash(meta_type::<StructFieldSwap>())
-        );
+/// # Panics
+///
+/// Panics if the runtime metadata version is not supported.
+///
+/// Supported versions: v14 and v15.
+pub fn metadata_to_latest(meta: RuntimeMetadataPrefixed) -> LatestRuntimeMetadata {
```

Comment:
with my suggestion of `pub struct RuntimeMetadata(RuntimeMetadataV15)` above, this could be a fn like `impl From<RuntimeMetadataV14> for RuntimeMetadata { ... }`.

In general, I think we should avoid panicking in this public function, so until the above sort of thing happens I'd prefer that this didn't panic and either took a specific RuntimeMetadataV14 or whatever or returned an `Option<RuntimeMetadataV15>` so we can panic where it's called instead rather than potentially forget about the panicking at the call site

Reply:
I've removed this after all since we were only using it in:
- codegen while extracting the V15 metadata
- metadata crate for unit tests

We could add it back later if need be when we have a better grasp of the metadata :D